migrator/data/DataImporter.java

```diff
@@ -361,97 +377,160 @@ private void insertEntity(TypeDB.Transaction transaction, DataProto.Item.Entity
+        private int instantiate(RelationType relationType, String originalID, List<DataProto.Item.Relation.Role> roleList) {
+            int rolesCreated = 0;
+            Relation relation = null;
+            for (DataProto.Item.Relation.Role roleMsg : roleList) {
+                RoleType roleType = getRoleType(relationType, roleMsg);
                 for (DataProto.Item.Relation.Role.Player playerMessage : roleMsg.getPlayerList()) {
-                    if (!isImported(playerMessage.getId())) return Optional.empty();
-                    else players.add(new Pair<>(roleType, getImportedThing(playerMessage.getId())));
+                    Thing player = getMappedThing(playerMessage.getId());
+                    if (player == null) continue;
+                    if (relation == null) {
+                        relation = relationType.create();
+                        recordMapping(relation.getIID(), originalID);
+                    }
+                    relation.addPlayer(roleType, player);
+                    rolesCreated++;
                 }
             }
-            return Optional.of(players);
+            return rolesCreated;
         }
+
+        private int tryExtend(Relation relation, DataProto.Item.Relation relationMsg) {
+            int rolesCreated = 0;
+            boolean stillIncomplete = false;
+            for (DataProto.Item.Relation.Role roleMsg : relationMsg.getRoleList()) {
+                RoleType roleType = getRoleType(relation.getType(), roleMsg);
+                for (DataProto.Item.Relation.Role.Player playerMessage : roleMsg.getPlayerList()) {
+                    Thing player = getMappedThing(playerMessage.getId());
+                    if (player == null) stillIncomplete = true;
+                    else if (!relation.getPlayers(roleType).findFirst(player).isPresent()) {
+                        relation.addPlayer(roleType, player);
+                        rolesCreated++;
+                    }
+                }
+            }
+            if (!stillIncomplete) recordCompletedID(relationMsg.getId());
+            status.roleCount.addAndGet(rolesCreated);
+            return rolesCreated;
+        }
+    }
+
+    private void importRelations() {
+        boolean progressMade;
+        do {
+            skippedRelations.set(false);
+            long before = status.relationCount.get() + status.roleCount.get();
+            new ParallelImport(Relations::new).executeImport();
+            progressMade = before < (status.relationCount.get() + status.roleCount.get());
+        } while (relationsUnfinished() && progressMade);
+
+        if (relationsUnfinished()) loadCyclicalRelations();
+        assert !conceptTracker.containsIncomplete();
     }
 
-    private void importSkippedRelations() {
+    private boolean relationsUnfinished() {
+        return skippedRelations.get() || conceptTracker.containsIncomplete();
+    }
+
+    private void loadCyclicalRelations() {
```

Comment:
we still have to handle the potential for relations that are pure cycles - eg. `r1 ---> r2` and `r2 ---> r1` at the same time - this requires we create the relation instances without any role players first, then attach the relation role players within that same transaction.